### DISCUSSION TITLE
how to visible column only admin ?

### DISCUSSION THREAD

**Discussion by shedenk** [filamentphp/filament#7302] [General]
how to visible column only admin ?

**Comment by danharrin** [comment]
`->visible(auth()->user()->isAdmin())`

**Reply by m-pernkopf** [reply]
Hy,

I know how to hide a column, but is there a way to create a column only for non admins? I am using a closure inside the column to get a specific state and show an icon (red/green) for the user, but the admin doesn't need that column (and it also seems to create performance issues, since the closure checks against an array, which is also created for non-admins, so not necessary if I am logged in as admin).

Regards, Michael

**Reply by danharrin** [reply]
It's the same example as I posted?

**Reply by m-pernkopf** [reply]
I understand, I already do that at the moment

Tables\Columns\IconColumn::make('isOnline')
                        ->getStateUsing(fn(Radcheck $record) => in_array($record->username, session('online_users')))->boolean()->label(__('tables.radcheck.status'))
                        ->visible(Auth::user()->isAdmin() === false)

but this only affects the visibility of the already existing column. I want to completely avoid creating the column if !isAdmin, like this:

        return $table
            ->columns([
                ((!$user->isAdmin()) ?
                    Tables\Columns\IconColumn::make('isOnline')
                        ->getStateUsing(fn(Radcheck $record) => in_array($record->username, session('online_users')))->boolean()->label(__('tables.radcheck.status'))
                ) 

but this doesn't work, and I don't see what I am missing here.

Regards, Michael

**Reply by 8888post** [reply]
```

        $columns = [
            Tables\Columns\TextColumn::make('first_name')
                ->required()
                ->maxLength(50),
            Tables\Columns\TextColumn::make('last_name')
                ->required()
                ->maxLength(50),
        ];


        $custom_columns = [
            Tables\Columns\IconColumn::make('isOnline')
                    ->getStateUsing(fn(Radcheck $record) => in_array($record->username, session('online_users')))
                    ->boolean()
                    ->label(__('tables.radcheck.status'))
        ];

        if(!$user->isAdmin(){
            foreach($custom_columns as $cc){
                $columns[] = $cc;
            }
        }

        return $table
            ->columns([
                    $columns
            ]);
```

**Reply by danharrin** [reply]
If the column is not visible, `getStateUsing()` will not be run, so there is no performance issue
